create lesson on testing concepts (#70)
Arcadia-Science/arcadia-computational-training: docs/arcadia-users-group/20231104-testing-concepts/lesson.md

* **Exceptions and assertions**: While writing code, `exceptions` and `assertions` can be added to sound an alarm as runtime problems come up. 
These kinds of tests, are embedded in the software iteself and handle, as their name implies, exceptional cases rather than the norm. 
* **Unit tests**: Unit tests investigate the behavior of units of code (such as functions, classes, or data structures).
By validating each software unit across the valid range of its input and output parameters, tracking down unexpected behavior that may appear when the units are combined is made vastly simpler.
* **Integration tests**: Integration tests check that various pieces of the software work together as expected. 
* **Regression tests**: Regression tests defend against new bugs, or regressions, which might appear due to new software and updates.

While each of these types of tests has a different definition, in practice there isn't always a firm delineation between each type.

## Assertions

Assertions are the simplest type of test.
They are used as a tool for bounding acceptable behavior during runtime.
The assert keyword in python has the following behavior:

```

> assert True == False

Reviewer: nit: this example may be confusing syntactically, consider something that looks more like a typical condition, such as 
```
a = 1
assert a > 2, "a is below the minimum value"
```

## Exceptions

Exceptions are more sophisticated than assertions.
When an error is encountered, an informative exception is 'thrown' or 'raised'.

For example, instead of the assertion in the case before, an exception can be used.

```
def mean(num_list):
    if len(num_list) == 0:
      raise Exception("The algebraic mean of an empty list is undefined. "
                      "Please provide a list of numbers")
    else:
      return sum(num_list)/len(num_list)
```

Once an exception is raised, it will be passed upward in the program scope.
An exception can be used to trigger additional error messages or an alternative behavior. 
Rather than immediately halting code execution, the exception can be 'caught' upstream with a try-except block.
When wrapped in a try-except block, the exception can be intercepted before it reaches global scope and halts execution.

To add information or replace the message before it is passed upstream, the try-catch block can be used to catch-and-reraise the exception:

```
def mean(num_list):
    try:
        return sum(num_list)/len(num_list)

>     except ZeroDivisionError as detail :

Author: ooh yes love that, much more clear, thank you!

        raise ZeroDivisionError(detail.__str__() + "\n" +  msg)
```

Alternatively, the exception can be handled apropriately for the use case.
If an alternative behavior is preferred, the exception can be disregarded and a responsive behavior can be implemented like so:

```
def mean(num_list):
    try:
        return sum(num_list)/len(num_list)
    except ZeroDivisionError :
        return 0
```

If a single function might raise more than one type of exception, each can be caught and handled separately.

```
def mean(num_list):
    try:
        return sum(num_list)/len(num_list)
    except ZeroDivisionError :
        return 0
    except TypeError as detail :
        msg = "The algebraic mean of an non-numerical list is undefined.\
               Please provide a list of numbers."
        raise TypeError(detail.__str__() + "\n" +  msg)
```

Exceptions have the advantage of being simple to include and when accompanied by useful help message, can be helpful to the user.
However, not all behaviors can or should be found with runtime exceptions.
Most behaviors should be validated with unit tests.

## Unit tests

> Unit tests are so called because they exercise the functionality of the code by interrogating individual functions and methods.

Author: ```suggestion
Unit tests are so called because they test the functionality of the code by interrogating individual functions and methods.
```

Later in this lesson, we'll use a test framework to collect and run them.
Using a test framework makes running tests streamlined.

Unit tests are typically made of three pieces, some set-up, a number of assertions, and some tear-down. 
Set-up can be as simple as initializing the input values or as complex as creating and initializing concrete instances of a class.
Ultimately, the test occurs when an assertion is made, comparing the observed and expected values.
For example, let us test that our mean function successfully calculates the known value for a simple list.

Before running the next code, save your `mean` function to a file called `mean.py` in the working directory.

You can use this code to save to file:

```
def mean(num_list):
    try:
        return sum(num_list)/len(num_list)
    except ZeroDivisionError :
        return 0
    except TypeError as detail :
        msg = "The algebraic mean of an non-numerical list is undefined.\
               Please provide a list of numbers."
        raise TypeError(detail.__str__() + "\n" +  msg)
```

Now, back in your Jupyter Notebook run the following code:

```
from mean import *

> def test_ints():

Author: ```suggestion
def test_mean_with_ints():
```

    num_list = [1, 2, 3, 4, 5]

>     obs = mean(num_list)
>     exp = 3
>     assert obs == exp

Author: ```suggestion
    observed_value = mean(num_list)
    expected_value = 3
    assert observed_value == expected_value
```

```

The test above:

* sets up the input parameters (the list `[1, 2, 3, 4, 5]`);
* collects the observed result;
* declares the expected result (calculated with our human brain);
* and compares the two with an assertion.

A unit test suite is made up of many tests just like this one.
A single implemented function may be tested in numerous ways.

In a file called `test_mean.py`, implement the following code:

```
from mean import *

> def test_ints():

Author: ```suggestion
def test_mean_with_ints():
```

    num_list = [1, 2, 3, 4, 5]

>     obs = mean(num_list)
>     exp = 3
>     assert obs == exp

Author: ```suggestion
    observed_value = mean(num_list)
    expected_value = 3
    assert observed_value == expected_value
```

> def test_zero():

Author: ```suggestion
def test_mean_with_zero():
```

    num_list=[0, 2, 4, 6]

>     obs = mean(num_list)
>     exp = 3
>     assert obs == exp

Author: ```suggestion
    observed_value = mean(num_list)
    expected_value = 3
    assert observed_value == expected_value
```

> def test_double():

Author: ```suggestion
def test_mean_with_double():
```

    # This one will fail in Python 2

>     num_list=[1, 2, 3, 4]
>     obs = mean(num_list)
>     exp = 2.5
>     assert obs == exp

Author: ```suggestion
    num_list=[1, 2, 3, 4]
    observed_value = mean(num_list)
    expected_value = 2.5
    assert observed_value == expected_value
```

> def test_long():

Author: ```suggestion
def test_mean_with_long():
```

    big = 100000000

>     obs = mean(range(1,big))
>     exp = big/2.0
>     assert obs == exp

Author: ```suggestion
    observed_value = mean(range(1,big))
    expected_value = big/2.0
    assert observed_value == expected_value
```

> def test_complex():

Author: ```suggestion
def test_mean_with_complex():
```

    # given that complex numbers are an unordered field
    # the arithmetic mean of complex numbers is meaningless
    num_list = [2 + 3j, 3 + 4j, -32 - 2j]

>     obs = mean(num_list)
>     exp = NotImplemented
>     assert obs == exp

Author: ```suggestion
    observed_value = mean(num_list)
    expected_value = NotImplemented
    assert observed_value == expected_value
```

```

Use Jupyter Notebook to import the `test_mean` package and run each test like this:

```
from test_mean import *

> test_ints()
> test_zero()
> test_double()
> test_long()
> test_complex()  ## Please note that this one might fail. You'll get an error message showing which tests failed

Author: ```suggestion
test_mean_with_ints()
test_mean_with_zero()
test_mean_with_double()
test_mean_with_long()
test_mean_with_complex()  ## Please note that this one might fail. You'll get an error message showing which tests failed
```

```

We just wrote and ran five tests for our `mean()` function~

> 

Reviewer: ```suggestion

You may have noticed that several of the tests look very similar to each other -- they introduce an input, call `mean()`, and test its output against an expected value. We'll come back to this later, offering a way to write a single test that applies to multiple inputs.
```

```

<details>
 <summary> Challenge solution </summary>

There are many ways this challenge could be solved.
One way is to check for the presence of complex numbers before calculating the mean.

```
def mean(num_list):
    if any(isinstance(num, complex) for num in num_list):
       return NotImplemented
    else:
        try:
            return sum(num_list)/len(num_list)
        except ZeroDivisionError :
            return 0
        except TypeError as detail :
            msg = "The algebraic mean of an non-numerical list is undefined.\
                   Please provide a list of numbers."
            raise TypeError(detail.__str__() + "\n" +  msg)
```

</details>

Note, using `pytest -v` will result in `pytest` listing which tests are executed and whether they pass or not:

```
pytest
```

```
collected 5 items

test_mean.py .....

test_mean.py::test_ints PASSED
test_mean.py::test_zero PASSED
test_mean.py::test_double PASSED
test_mean.py::test_long PASSED
test_mean.py::test_complex PASSED

========================== 5 passed in 2.57 seconds ===========================
```

As we write more code, we would write more tests, and pytest would produce more dots.

> 

Reviewer: ```suggestion

### Parametrized tests

Often, different unit tests of the same "unit" have similar logic, but are applied to different inputs; this is the case with our `mean()` test suite above. `pytest` offers a convenient mechanism to write just a single test and apply it to multiple inputs, called test parameterization. This way, adding a new test case is as simple as defining one more input and expected output.

To parametrize a test, we "decorate" it with its expected inputs and outputs, and `pytest` will expand the test to run as many times as needed to check all inputs. In the `pytest` output, this will look exactly like running multiple tests that have different functions.

<details>
<summary> Python decorators </summary>
"Decorators" are a Python way to enhance or wrap a function with additional behavior. They are added just before a function definition, using a name that starts with `@`, and can take arguments separate from the function arguments, for example:

\@log_to_file("output.txt")
def mean(num_list):
  ...

</details>

For example, here's a parametrized version of our test suite for `mean()`. In this case, we decorate the test with two parameters that can be used within the test: `num_list` and `expected_mean`. The first parameter will be used to call the `mean()` function, and the second is its expected result, which the test will validate. We then provide a list of matching pairs of these parameters, each of which will run as a test:

\```
@pytest.mark.parametrize("num_list,exp", [
    ([1,2,3,4,5], 3),
    ([0,2,4,6], 3),
    (range(1,10000), 10000/2.0),
  ])
  def test_mean(num_list, exp):
    obs = mean(num_list)
    assert obs == exp
\```

```

## Integration tests

Integration tests focus on gluing code together or the results of code when multiple functions are used.
See below for an conceptual example of an integration test.

Consider three functions `a()`, `b()`, and `c()` as a simplistic example.

> Function `a()` increments a number by one, `b()` multiplies a number by two, and `c()` composes them as defined below:

Reviewer 2: another nit: I would favor renaming these functions with more explicit names both for clarity and to avoid implicitly sanctioning the use of ambiguous/meaningless single-letter names

Author: oh i totally agree, good catch thank you!!

### GitHub Actions

[GitHub Actions](https://github.com/features/actions) is a continuous integration service provided by GitHub.
It's integrated directly into GitHub repositories and does not require additional accounts or external services.

To use GitHub Actions, create a directory called `.github` and within it, create another directory called `workflows`.

```
mkdir -p .github/workflows
```

Inside the `workflows` directory, you can create a YAML file (e.g. `ci.yml`) to define your continuous integration process:

```
name: Continuous Integration

on: [push, pull_request]

jobs:
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest, windows-latest]
        python-version: [3.7, 3.8, 3.9, '3.10', '3.11'] 

    steps:
    - uses: actions/checkout@v2
    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v2
      with:
        python-version: ${{ matrix.python-version }}
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install -r requirements.txt
    - name: Run tests

>       run: pytest

Reviewer 2: given that this is an intro to testing, and also that github actions are a bit opaque to begin with, it might be worth using a simpler action here without the matrix stuff, e.g. 

```
jobs:
  build:
    runs-on: ubuntu-latest 
    steps:
    - uses: actions/checkout@v2
    - name: Set up Python 3.10 
      uses: actions/setup-python@v2
      with:
        python-version: 3.10
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install -r requirements.txt
    - name: Run tests
      run: pytest
```

Reviewer 3: I agree with this. Or explain to them what all the matrix stuff does. 

We should also explain at the end that while this specific example uses `requirements.txt` to install dependencies, it can technically be installed in other ways as well (including conda)

# Testing concepts and terminology

> Note that this lesson has been modified from [The Carpentries Incubator](https://github.com/carpentries-incubator) lesson on [Python Testing](https://carpentries-incubator.github.io/python-testing/).
> Parts are reproduced in full, but the major changes were included to shorten the lesson to 60 minutes.

Everyone tests their software to some extent, if only by running it and trying it out.
Most programmers do a certain amount of exploratory testing while they write new code or change old code.

> This may involve printing or plotting outputs of your code as you program.

Reviewer 3: ```suggestion
This may involve printing or plotting the outputs of your code as you program.
```

> Systematic testing codifies these behaviors, allowing them to be automatically applied quickly and repeatedly over entire code bases.

Reviewer 3: I think in the introduction we should make it clearer why to test (making sure your function or program does what it is meant to do and handles different cases, making sure a change you introduce doesn't break things etc)

Author: oh good call. i modified the intro to add this, thank you!!

This is more overhead than we typically strive for in a lesson, but we hope that it's a chance to practice these skills to achieve a new goal.
In the future, we may provide a [GitPod](https://www.gitpod.io/) environment for learners to use while working through this lesson, however if possible we would prefer to empower users to start implementing tests on their own computers using their own setup.
 
To start this lesson, we'll begin by creating a conda environment that has the tools we'll need.

```
mamba env create -n augtest jupyter pytest 
conda activate augtest
```

We'll also create a folder to help us stay organized.
```
mkdir aug-mean
cd aug-mean
```

Once our environment is activated, start a Jupyter notebook

```
jupyter notebook
```

Now we can start learning about testing!

## An introduction to testing concepts

> There are many ways to test software, such as assertions, exceptions, unit tests, integration tests, and regresson tests.

Reviewer 3: ```suggestion
There are many ways to test software, such as assertions, exceptions, unit tests, integration tests, and regression tests.
```

* **Exceptions and assertions**: While writing code, `exceptions` and `assertions` can be added to sound an alarm as runtime problems come up. 

> These kinds of tests, are embedded in the software iteself and handle, as their name implies, exceptional cases rather than the norm. 

Reviewer 3: ```suggestion
These kinds of tests, are embedded in the software itself and handle, as their name implies, exceptional cases rather than the norm. 
```

* **Unit tests**: Unit tests investigate the behavior of units of code (such as functions, classes, or data structures).
By validating each software unit across the valid range of its input and output parameters, tracking down unexpected behavior that may appear when the units are combined is made vastly simpler.

> * **Integration tests**: Integration tests check that various pieces of the software work together as expected. 

Reviewer 3: I'd give a couple examples for the definitions of unit tests vs integration tests. 

Unit tests: a function, an individual Snakemake rule, a process in Nextflow, a cell in Jupyter
Integration tests: a set of Snakemake/Nextflow rules (or the full pipeline), the full Jupyter notebook

Author: nice -- added!

For example, instead of the assertion in the case before, an exception can be used.

```
def mean(num_list):
    if len(num_list) == 0:
      raise Exception("The algebraic mean of an empty list is undefined. "
                      "Please provide a list of numbers")
    else:
      return sum(num_list)/len(num_list)
```

Once an exception is raised, it will be passed upward in the program scope.
An exception can be used to trigger additional error messages or an alternative behavior. 
Rather than immediately halting code execution, the exception can be 'caught' upstream with a try-except block.
When wrapped in a try-except block, the exception can be intercepted before it reaches global scope and halts execution.

To add information or replace the message before it is passed upstream, the try-catch block can be used to catch-and-reraise the exception:

```
def mean(num_list):
    try:
        return sum(num_list)/len(num_list)
    except ZeroDivisionError as detail :
        msg = "The algebraic mean of an empty list is undefined. Please provide a list of numbers."
        raise ZeroDivisionError(detail.__str__() + "\n" +  msg)
```

> Alternatively, the exception can be handled apropriately for the use case.

Reviewer 3: ```suggestion
Alternatively, the exception can be handled appropriately for the use case.
```

    num_list = [2 + 3j, 3 + 4j, -32 - 2j]
    observed_value = mean(num_list)
    expected_value = NotImplemented
    assert observed_value == expected_value
```

Use Jupyter Notebook to import the `test_mean` package and run each test like this:

```
from test_mean import *

test_mean_with_ints()
test_mean_with_zero()
test_mean_with_double()
test_mean_with_long()
test_mean_with_complex()  ## Please note that this one might fail. You'll get an error message showing which tests failed
```

We just wrote and ran five tests for our `mean()` function~

You may have noticed that several of the tests look very similar to each other -- they introduce an input, call `mean()`, and test its output against an expected value. We'll come back to this later, offering a way to write a single test that applies to multiple inputs.
## Using the test framework `pytest`

We created a suite of tests for our mean function, but it was annoying to run them one at a time.
It would be a lot better if there were some way to run them all at once, just reporting which tests fail and which succeed.

Thankfully, that exists.
Recall our tests:

```
from mean import *

> def test_ints():
>     num_list = [1,2,3,4,5]
>     obs = mean(num_list)
>     exp = 3
>     assert obs == exp
> 
> def test_zero():
>     num_list=[0,2,4,6]
>     obs = mean(num_list)
>     exp = 3
>     assert obs == exp
> 
> def test_double():
>     # This one will fail in Python 2
>     num_list=[1,2,3,4]
>     obs = mean(num_list)
>     exp = 2.5
>     assert obs == exp
> 
> def test_long():
>     big = 100000000
>     obs = mean(range(1,big))
>     exp = big/2.0
>     assert obs == exp
> 
> def test_complex():
>     # given that complex numbers are an unordered field
>     # the arithmetic mean of complex numbers is meaningless
>     num_list = [2 + 3j, 3 + 4j, -32 - 2j]
>     obs = mean(num_list)
>     exp = NotImplemented
>     assert obs == exp

Author: ```suggestion
def test_mean_with_ints():
    num_list = [1, 2, 3, 4, 5]
    observed_value = mean(num_list)
    expected_value = 3
    assert observed_value == expected_value

def test_mean_with_zero():
    num_list=[0, 2, 4, 6]
    observed_value = mean(num_list)
    expected_value = 3
    assert observed_value == expected_value

def test_mean_with_double():
    # This one will fail in Python 2
    num_list=[1, 2, 3, 4]
    observed_value = mean(num_list)
    expected_value = 2.5
    assert observed_value == expected_value

def test_mean_with_long():
    big = 100000000
    observed_value = mean(range(1,big))
    expected_value = big/2.0
    assert observed_value == expected_value

def test_mean_with_complex():
    # given that complex numbers are an unordered field
    # the arithmetic mean of complex numbers is meaningless
    num_list = [2 + 3j, 3 + 4j, -32 - 2j]
    observed_value = mean(num_list)
    expected_value = NotImplemented
    assert observed_value == expected_value
```

```

Once these tests are written in a file called `test_mean.py`, the command `pytest` can be run on the terminal or command line from the directory containing the tests (note that you'll have to use `py.test` for older versions of the `pytest` package):

```
pytest
```

```
collected 5 items

test_mean.py ....F

================================== FAILURES ===================================
________________________________ test_complex _________________________________

>     def test_complex():
>         # given that complex numbers are an unordered field
>         # the arithmetic mean of complex numbers is meaningless
>         num_list = [2 + 3j, 3 + 4j, -32 - 2j]
>         obs = mean(num_list)
>         exp = NotImplemented
> >       assert obs == exp

Author: ```suggestion
    def test_mean_with_complex():
        # given that complex numbers are an unordered field
        # the arithmetic mean of complex numbers is meaningless
        num_list = [2 + 3j, 3 + 4j, -32 - 2j]
        observed_value = mean(num_list)
        expected_value = NotImplemented
>       assert observed_value == expected_value
```

Without changing the tests, alter the mean.py file from the previous section until it passes.
When it passes, `pytest` will produce results like the following:

```
pytest
```

```
collected 5 items

test_mean.py .....

========================== 5 passed in 2.68 seconds ===========================
```

<details>
 <summary> Challenge solution </summary>

There are many ways this challenge could be solved.
One way is to check for the presence of complex numbers before calculating the mean.

```
def mean(num_list):
    if any(isinstance(num, complex) for num in num_list):
       return NotImplemented
    else:
        try:
            return sum(num_list)/len(num_list)
        except ZeroDivisionError :
            return 0
        except TypeError as detail :
            msg = "The algebraic mean of an non-numerical list is undefined.\
                   Please provide a list of numbers."
            raise TypeError(detail.__str__() + "\n" +  msg)
```

</details>

Note, using `pytest -v` will result in `pytest` listing which tests are executed and whether they pass or not:

```
pytest
```

```
collected 5 items

test_mean.py .....

> test_mean.py::test_ints PASSED
> test_mean.py::test_zero PASSED
> test_mean.py::test_double PASSED
> test_mean.py::test_long PASSED
> test_mean.py::test_complex PASSED

Author: ```suggestion
test_mean.py::test_mean_with_ints PASSED
test_mean.py::test_mean_with_zero PASSED
test_mean.py::test_mean_with_double PASSED
test_mean.py::test_mean_with_long PASSED
test_mean.py::test_mean_with_complex PASSED
```

The discrepancy between current and previous code outputs is termed a regression.

Like integration tests, regression tests are high-level and often encompass the entire code base.
A prevalent regression test strategy extends across multiple code versions.
For instance, an input file for version X of a workflow is processed, and the output file is saved, typically online.
While developing version Y, the test suite automatically fetches the output for version X, processes the same input file for version Y, and contrasts the two output files.
Any significant discrepancies trigger a test failure.
Regression tests can identify failures missed by integration and unit tests.
Each project may adopt a slightly varied approach to regression testing, based on its software requirements.
Testing frameworks aid in constructing regression tests but don’t provide additional sophistication beyond the discussed concepts.

## Continuous integration

Continuous integration makes running tests as easy as possible by integrating the test suite into the development process. 
Every time a change is made to the repository, the continuous integration system builds and checks that code.

> Based on instructions you provide, a continuous integration server can:

Reviewer 3: ```suggestion
Based on the instructions you provide, a continuous integration server can:
```

- spin up those instances with different software versions (i.e. python 2.7 and python 3.0)
- run the build and test scripts
- check for errors
- report the results.

Since the first step the server conducts is to check out the code from a repository, we'll need to put our code online to make use of this kind of server.

### Set Up a Mean Git Repository on GitHub

Our `mean.py` `test_mean.py` files can be the contents of a repository on GitHub.

* Go to GitHub and [create a repository](https://github.com/new) called aug-mean. Do this in your own user account and don't add any files to the new repo (no README/LICENSE, etc.).
* Turn the `aug-mean` directory that we've been working in on your computer into a git repository following the "…or create a new repository on the command line" instructions on GitHub:
```
echo "# aug-mean" >> README.md
git init
git add README.md
git commit -m "first commit"
git branch -M main
git remote add origin git@github.com:yourusername/aug-mean.git
git push -u origin main
```
* Create a new branch (`git checkout -b yourinitials/init`).
* Use git to `add`, `commit`, and `push` the two files `mean.py` and `test_mean.py` to GitHub.

### GitHub Actions

> [GitHub Actions](https://github.com/features/actions) is a continuous integration service provided by GitHub.

Reviewer 3: We should make sure to mention this is not free and that you get free credits.

Author: added

It's integrated directly into GitHub repositories and does not require additional accounts or external services.

To use GitHub Actions, create a directory called `.github` and within it, create another directory called `workflows`.

```
mkdir -p .github/workflows
```

Inside the `workflows` directory, you can create a YAML file (e.g. `ci.yml`) to define your continuous integration process:

```

> name: Continuous Integration

Reviewer 3: I'd probably name it more testing focused

Author: fixed

> on: [push, pull_request]

Reviewer 3: I'd also make sure you explain what this does

Author: done

## An introduction to testing concepts

There are many ways to test software, such as assertions, exceptions, unit tests, integration tests, and regression tests.

* **Exceptions and assertions**: While writing code, `exceptions` and `assertions` can be added to sound an alarm as runtime problems come up. 
These kinds of tests, are embedded in the software itself and handle, as their name implies, exceptional cases rather than the norm. 
* **Unit tests**: Unit tests investigate the behavior of units of code (such as functions, classes, or data structures).
By validating each software unit across the valid range of its input and output parameters, tracking down unexpected behavior that may appear when the units are combined is made vastly simpler.
Some examples of things a unit test might test include functions, individual Snakemake rules, a process in Nextflow, or a cell in a Jupyter notebook.
* **Integration tests**: Integration tests check that various pieces of the software work together as expected.
Some examples of things an integration test might test include a set of Snakemake rule or Nextflow processes or the execution of an entire Jupyter notebook.

> * **Regression tests**: Regression tests defend against new bugs, or regressions, which might appear due to new software and updates. Regression tests can also refer to test for decreases in performance (run time, memory usage, etc.) or in the qualit of some output (the resolution of a rendered graph, accuracy of a set of predictions, etc.).

Reviewer 2: ```suggestion
* **Regression tests**: Regression tests defend against new bugs, or regressions, which might appear due to new software and updates. Regression tests can also refer to tests for decreases in performance (run time, memory usage, etc.) or in the quality of some output (the resolution of a rendered graph, accuracy of a set of predictions, etc.).
```

While each of these types of tests has a different definition, in practice there isn't always a firm delineation between each type.

## Assertions

Assertions are the simplest type of test.
They are used as a tool for bounding acceptable behavior during runtime.
The assert keyword in python has the following behavior:

```
assert True == False
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  AssertionError
```

```
assert True == True
```

> ```

Reviewer: nit: rm

    # the arithmetic mean of complex numbers is meaningless
    num_list = [2 + 3j, 3 + 4j, -32 - 2j]
    observed_value = mean(num_list)
    expected_value = NotImplemented
    assert observed_value == expected_value
```

Use Jupyter Notebook to import the `test_mean` package and run each test like this:

```
from test_mean import *

test_mean_with_ints()
test_mean_with_zero()
test_mean_with_double()
test_mean_with_long()
test_mean_with_complex()  ## Please note that this one might fail. You'll get an error message showing which tests failed
```

We just wrote and ran five tests for our `mean()` function~

You may have noticed that several of the tests look very similar to each other -- they introduce an input, call `mean()`, and test its output against an expected value. We'll come back to this later, offering a way to write a single test that applies to multiple inputs.
## Using the test framework `pytest`

We created a suite of tests for our mean function, but it was annoying to run them one at a time.
It would be a lot better if there were some way to run them all at once, just reporting which tests fail and which succeed.

Thankfully, that exists.
Recall our tests:

> ```

Reviewer: nit: use "```python" to turn on syntax highlighting for longer blobs, for readability.

Reviewer 2: +1 almost mentioned this but felt like I'd spent my nit budget on all those nomenclature changes 😛

Author: there is no nit budget! i didn't actually know about syntax highlighting in markdown that way! I'll go through and add it, thanks guys!

    num_list = [2 + 3j, 3 + 4j, -32 - 2j]
    observed_value = mean(num_list)
    expected_value = NotImplemented
    assert observed_value == expected_value
```

Once these tests are written in a file called `test_mean.py`, the command `pytest` can be run on the terminal or command line from the directory containing the tests (note that you'll have to use `py.test` for older versions of the `pytest` package):

```
pytest
```

```
collected 5 items

test_mean.py ....F

================================== FAILURES ===================================
________________________________ test_complex _________________________________

    def test_mean_with_complex():
        # given that complex numbers are an unordered field
        # the arithmetic mean of complex numbers is meaningless
        num_list = [2 + 3j, 3 + 4j, -32 - 2j]
        observed_value = mean(num_list)
        expected_value = NotImplemented
>       assert observed_value == expected_value
E       assert (-9+1.6666666666666667j) == NotImplemented

test_mean.py:34: AssertionError
===================== 1 failed, 4 passed in 2.71 seconds ======================
```

> In the above case, the pytest package sniffed out the tests in the directory and ran them together to produce a report of the sum of the files and functions matching the regular expression `[Tt]est[-_]*`.

Reviewer: super nit: the regex is missing a period before the asterisk.

> In the above case, the pytest package sniffed out the tests in the directory and ran them together to produce a report of the sum of the files and functions matching the regular expression `[Tt]est[-_]*`.

Author: ```suggestion
In the above case, the pytest package sniffed out the tests in the directory and ran them together to produce a report of the sum of the files and functions matching the regular expression `[Tt]est[-_].*`.
```

not a nit!! an important and good catch, thank you!

```

As we write more code, we would write more tests, and pytest would produce more dots.

### Parametrized tests

Often, different unit tests of the same "unit" have similar logic, but are applied to different inputs; this is the case with our `mean()` test suite above.
`pytest` offers a convenient mechanism to write just a single test and apply it to multiple inputs, called test parameterization.
This way, adding a new test case is as simple as defining one more input and expected output.

To parametrize a test, we "decorate" it with its expected inputs and outputs, and `pytest` will expand the test to run as many times as needed to check all inputs.
In the `pytest` output, this will look exactly like running multiple tests that have different functions.

<details>
<summary> Python decorators </summary>
"Decorators" are a Python way to enhance or wrap a function with additional behavior.
They are added just before a function definition, using a name that starts with `@`, and can take arguments separate from the function arguments, for example:

<pre><code>
@log_to_file("output.txt")
def mean(num_list):
  ...
</code></pre>

> </details>

Reviewer 2: This might be too much but I think it would help to demystify decorators to show explicitly what "wrapping" means:

```suggestion
This code is equivalent to wrapping or "decorating" the `mean` function with the function `log_to_file("output.txt")`:
<pre><code>
def _mean(num_list):
    ...

mean = log_to_file("output.txt")(_mean)
</code></pre>
</details>
The ampersand-based "decorator" syntax is just a clearer and more readable way of wrapping a function with another function. In this example, that other function is in fact `log_to_file("output.xt")`. How is this possible? `log_to_file` is a function that takes a filename as an input and _returns another function_ that itself takes a function as an argument and then returns a new "wrapped" function. 
```

For example, here's a parametrized version of our test suite for `mean()`.

> In this case, we decorate the test with two parameters that can be used within the test: `num_list` and `expected_mean`.

Reviewer: looks like the name below is `expected_value`.

> In this case, we decorate the test with two parameters that can be used within the test: `num_list` and `expected_mean`.

Author: ```suggestion
In this case, we decorate the test with two parameters that can be used within the test: `num_list` and `expected_value`.
```

* `jobs`: Specifies what the workflow will actually run. In this case, we specify that we want to run on the latest version of the ubuntu operating system using python version 3.10. These instructions are enough to launch a computer with python running on it. Then, we specify that we want to install dependencies from a `requirements.txt` file using pip. Lastly, we run our tests using `pytest`.

<details>
<summary> Running GitHub Actions workflows on multiple operating systems using a matrix</summary>

Often times, developers want to check that their tests will pass not just with an ubuntu operating system and one version of Python, but with many operating systems and many versions of Python.
This can be done using a matrix, which we demonstrate below.
We specify lists of operating systems and python versions that we want to run our CI with, and then use a matrix call to run them all.

```
name: pytest CI

on: [push, pull_request]

jobs:
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest, windows-latest]

>         python-version: [3.7, 3.8, 3.9, '3.10', '3.11'] 

Reviewer 2: ```suggestion
        python-version: [3.7, 3.8, 3.9, 3.10, 3.11] 
```

Author: these ones need to be in quotes -- it fails by testing for `3.1` otherwise. it's very weird syntax tho...

Reviewer 2: yikes. not an excuse but I checked this w gpt4 and it got it wrong too. maybe they should all be quoted, just for consistency?

Author: woot tested quotes on everything and that works, so added!

    steps:
    - uses: actions/checkout@v2
    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v2
      with:
        python-version: ${{ matrix.python-version }}
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install -r requirements.txt
    - name: Run tests
      run: pytest
```

</details>

### Triggering CI

1. Add `.github/workflows/ci.yml` to your repository
2. Commit and push it.
3. Open a pull request with your changes.
4. Check how the CI is going in the PR.

Right now, the CI is failing.
Why is that?
When we set up for today's lesson, we used conda to install pytest.
In the CI workflow, we specified that our dependencies (in this case pytest) should be installed from a `requirements.txt` file.
That `requirements.txt` file is a conventional way to list all of the python packages that we need.
We haven't created that file yet.
Let's go ahead an create it, add it, commit it, and push it.
Since we need pytest, the `requirements.txt` file looks like this:

```
pytest==7.4.3
```

> Pushing new changes to our branch automatically re-triggers the CI teststo re-run.

Reviewer 2: ```suggestion
Pushing new changes to our branch automatically re-triggers the CI tests to re-run.
```
